### ISSUE TITLE
PCSs should support different opening locations for each matrix

### ISSUE THREAD

**Issue by dlubarov** [Plonky3/Plonky3#187]
Currently we have

```
    fn open_multi_batches(
        &self,
        prover_data_and_points: &[(&Self::ProverData, &[EF])],
        challenger: &mut Challenger,
    ) -> (OpenedValues<EF>, Self::Proof);
```

Each `ProverData` corresponds to a batch of (potentially multiple) committed matrices, but each such batch can only have one set of opening points, `&[EF]`.

In typical usage, the opening locations within a batch are almost uniform, like all "round 1" traces are typically opened at `{zeta, g * zeta}`. However, the shift `g` depends on trace length, and within a batch there may be traces of different lengths.

So I think we need to generalize it to something like

```
prover_data_and_points: &[(&Self::ProverData, &[Vec<EF>])],
```

This isn't an issue for uni-stark, where each batch contains a single matrix, but it's a blocker for the Valida prover integration (cc @maxgillett).

**Comment by npwardberkeley** [comment]
Regarding the openings the function produces, does this mean that `OpenedValuesForPoint<F>` should change from `Vec<OpenedValuesForMatrix<F>>` to `Vec<Vec<OpenedValuesForMatrix<F>>>`?

**Comment by npwardberkeley** [comment]
Also, how do you feel about aliasing `Vec<EF>` to `OpeningSet` or something like that? I think it'll help to have an additional layer of abstraction because the vec-of-array-of-vecs can get hard to follow at certain points. Or is there a better name for it?

**Comment by rjeli** [comment]
Maybe it's worth creating an actual struct with convenience methods, like
```rust
prover_data_and_points: &[(&Self::ProverData, &BatchOpeningSet)],
```
or
```rust
prover_data: &[&Self::ProverData],
points: &BatchOpeningSet,
```
(depending on whether it ends up better to have one per `Data` or one for all `Data`). Then you would have `BatchOpening::opened_points_for_matrix(batch_idx, mat_idx)` etc

I was already getting my eyes crossed working on `LdtBasedPcs`, and it's not even that complicated:
https://github.com/Plonky3/Plonky3/blob/844d85f09337608708abd3296b89d5193117ee00/ldt/src/ldt_based_pcs.rs#L165-L191

**Comment by jorgeantonio21** [comment]
@npwardberkeley I made refactors a few refactors in #194 which included changing the `OpenedValuesForMatrix<F>` to consist of a vector of `OpenedValuesForPoints<F>`. I think this change makes it more consistent with the goals of this issue to work with a struct that represents matrix openings at several points.

@rjeli I like your suggestion, I can add something along those lines in the mentioned PR.

**Comment by dlubarov** [comment]
Done by @jorgeantonio21
